### PR TITLE
Add aov

### PR DESCRIPTION
**Pull request by AlexanderVeselov**
- Background
- Opacity
- Object ID
- Group ID

### CODE REVIEW

File: Baikal/Controllers/clw_scene_controller.cpp [hunk 1]
```diff
@@ -386,6 +386,7 @@ namespace Baikal
         std::size_t num_uvs_written = 0;
         std::size_t num_indices_written = 0;
         std::size_t num_shapes_written = 0;
+        std::size_t num_shapes_additional_written = 0;
```
Comment (Kvalme) [inline Baikal/Controllers/clw_scene_controller.cpp]: I believe that you can reuse num_shapes_written to avoid possibility of inconsistent state between Shapes and ShapeAdditionData

File: Baikal/Estimators/path_tracing_estimator.cpp [hunk 2]
```diff
@@ -375,7 +391,7 @@ namespace Baikal
         shadekernel.SetArg(argc++, m_render_data->rays[(pass + 1) & 0x1]);
         shadekernel.SetArg(argc++, output);
         shadekernel.SetArg(argc++, scene.input_map_data);
-
+
```
Comment (Kvalme) [inline Baikal/Estimators/path_tracing_estimator.cpp]: unneeded change

File: Baikal/Kernels/CL/fill_aovs_uberv2.cl [hunk 3]
```diff
@@ -154,6 +168,27 @@ KERNEL void FillAOVsUberV2(
         if (shape_ids_enabled)
             aov_shape_ids[idx].x = -1;
 
+        if (background_enabled)
+        {
+            if (env_light_idx != -1)
```
Comment (Kvalme) [inline Baikal/Kernels/CL/fill_aovs_uberv2.cl]: Please, double check priority. What renders first in color aov env_light or background override?

File: Baikal/Kernels/CL/fill_aovs_uberv2.cl [hunk 4]
```diff
@@ -338,9 +373,24 @@ KERNEL void FillAOVsUberV2(
             
             if (mesh_id_enabled)
             {
-                mesh_id[idx] = make_float4(isect.shapeid, isect.shapeid, isect.shapeid, 1.f);
+                Sampler shapeid_sampler;
```
Comment (Kvalme) [inline Baikal/Kernels/CL/fill_aovs_uberv2.cl:373]: What's happening here?

Comment (Kvalme) [reply Baikal/Kernels/CL/fill_aovs_uberv2.cl:373]: You need to simply save shape_id into aov.

File: Baikal/Kernels/CL/path.cl [hunk 5]
```diff
@@ -39,7 +39,8 @@ typedef enum _PathFlags
 {
     kNone = 0x0,
     kKilled = 0x1,
-    kScattered = 0x2
+    kScattered = 0x2,
+    kContainsOpacity = 0x4
```
Comment (Kvalme) [inline Baikal/Kernels/CL/path.cl]: kOpaque?

File: Baikal/Kernels/CL/path_tracing_estimator.cl [hunk 6]
```diff
@@ -242,11 +276,11 @@ KERNEL void FilterPathStream(
 
         if (Path_IsAlive(path))
         {
-            bool kill = (length(Path_GetThroughput(path)) < CRAZY_LOW_THROUGHPUT);
+            bool kill = (length(Path_GetThroughput(path)) < CRAZY_LOW_THROUGHPUT) || (isects[global_id].shapeid < 0);
```
Comment (Kvalme) [inline Baikal/Kernels/CL/path_tracing_estimator.cl:279]: You changed kill-logic here. Please, describe why it was changed.

File: Baikal/Renderers/renderer.h [hunk 7]
```diff
@@ -51,7 +51,14 @@ namespace Baikal
     public:
         enum class OutputType
         {
+            // Multi-pass outputs that will
+            // be rendered in Estimator kernel
             kColor = 0,
+            kOpacity,
+            kVisibility,
+            kMaxMultiPassOutput,
```
Comment (Kvalme) [inline Baikal/Renderers/renderer.h:59]: That's bad. Possibly it should be kMaxMultiPassOutput = kVisibility

File: BaikalIO/scene_test_io.cpp [hunk 8]
```diff
@@ -263,6 +263,14 @@ namespace Baikal
                                     , false);
             scene->AttachShape(floor);
 
+            auto mat = UberV2Material::Create();
```
Comment (Kvalme) [inline BaikalIO/scene_test_io.cpp]: Changes in this file seems to be unrelated to this CR

File: BaikalStandalone/Application/application.cpp [hunk 9]
```diff
@@ -758,19 +758,37 @@ namespace Baikal
...
-            "Geometric normal\0"
-            "Texture coords\0"
-            "Wire\0"
-            "Albedo\0"
-            "Tangent\0"
-            "Bitangent\0"
-            "Gloss\0"
-            "Depth\0\0"
-            ;
+        static const std::vector<std::pair<Baikal::Renderer::OutputType, char const*>> kBaikalOutputs =
```
Comment (Kvalme) [inline BaikalStandalone/Application/application.cpp:753]: maybe replace it with std::map and then you willn't need to made second vector

File: BaikalStandalone/Application/cl_render.cpp [hunk 10]
```diff
@@ -605,8 +607,17 @@ namespace Baikal
     {
         for (std::size_t i = 0; i < m_cfgs.size(); ++i)
         {
+            // Not good
+            if (m_output_type == Renderer::OutputType::kOpacity || m_output_type == Renderer::OutputType::kVisibility)
+            {
+                m_cfgs[i].renderer->SetOutput(Renderer::OutputType::kColor, nullptr);
```
Comment (Kvalme) [inline BaikalStandalone/Application/cl_render.cpp]: Why you made it empty here and dummy in next if?

File: BaikalStandalone/Application/cl_render.h [hunk 11]
```diff
@@ -115,6 +115,7 @@ namespace Baikal
         std::promise<int> m_promise;
         bool m_shape_id_requested = false;
         OutputData m_shape_id_data;
+        OutputData m_dummy_output_data;
```
Comment (Kvalme) [inline BaikalStandalone/Application/cl_render.h:122]: What this for?

File: BaikalTest/light.h [hunk 12]
```diff
@@ -1034,3 +1034,51 @@ TEST_F(LightTest, Light_IblTransparencyOverride)
         ASSERT_TRUE(CompareToReference(oss.str()));
     }
 }
+
+TEST_F(LightTest, Light_IblDifferentOverrides)
```
Comment (Kvalme) [inline BaikalTest/light.h]: Is it related to current CR?

File: CMakeLists.txt [hunk 13]
```diff
@@ -88,11 +88,9 @@ endif (BAIKAL_ENABLE_STANDALONE)
 if (BAIKAL_ENABLE_TESTS)
     add_subdirectory(Gtest)
     add_subdirectory(BaikalTest)
-# Temporarily remove RprTest from build
-# it will be rewrited in gtest
-#    if (BAIKAL_ENABLE_RPR)
-#        add_subdirectory(RprTest)
-#    endif (BAIKAL_ENABLE_RPR)
+    if (BAIKAL_ENABLE_RPR)
```
Comment (Kvalme) [inline CMakeLists.txt]: This should be done in previous CR

**Review by Kvalme — CHANGES_REQUESTED** [review]
Will continue review after RprTest merge

File: Baikal/Kernels/CL/path_tracing_estimator.cl (hunk 6 above)
Comment (AlexanderVeselov) [reply Baikal/Kernels/CL/path_tracing_estimator.cl:279]: If we don't kill them, we'll skip missed paths in GatherOpacity kernel, because there we consider killed paths.

File: BaikalStandalone/Application/cl_render.h (hunk 11 above)
Comment (AlexanderVeselov) [reply BaikalStandalone/Application/cl_render.h:122]: Opacity and visibility outputs require color output to be set, but we don't have another one viewport to render color there, so we use a "dummy" output

File: Baikal/Kernels/CL/fill_aovs_uberv2.cl (hunk 4 above)
Comment (AlexanderVeselov) [reply Baikal/Kernels/CL/fill_aovs_uberv2.cl:373]: Why? We've agreed to use random color in mesh id aov, haven't we?

File: Baikal/Kernels/CL/fill_aovs_uberv2.cl [hunk 14]
```diff
@@ -124,8 +138,8 @@ KERNEL void FillAOVsUberV2(
     // And negative number if there is no any shape in the pixel
     GLOBAL float4* restrict aov_shape_ids,
     // NOTE: following are fake parameters, handled outside
-    int visibility_enabled,
-    GLOBAL float4* restrict aov_visibility,
+    //int visibility_enabled,
```
Comment (Kvalme) [inline Baikal/Kernels/CL/fill_aovs_uberv2.cl]: please, remove

**Review by Kvalme — CHANGES_REQUESTED** [review]
